sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetInteroperabilitySuite.scala

```diff
@@ -87,4 +96,113 @@ class ParquetInteroperabilitySuite extends ParquetCompatibilityTest with SharedS
+              (SQLConf.PARQUET_INT96_TIMESTAMP_CONVERSION.key, applyConversion.toString()),
+              (SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key, vectorized.toString())
+          ) {
+            val read = spark.read.parquet(tableDir.getAbsolutePath).collect()
+            assert(read.size === 6)
+            // if we apply the conversion, we'll get the "right" values, as saved by impala in the
+            // original file.  Otherwise, they're off by the local timezone offset, set to
+            // America/Los_Angeles in tests
+            val impalaExpectations = if (applyConversion) {
+              impalaFileData
+            } else {
+              impalaFileData.map { ts =>
+                DateTimeUtils.toJavaTimestamp(DateTimeUtils.convertTz(
+                  DateTimeUtils.fromJavaTimestamp(ts),
+                  TimeZone.getTimeZone("UTC"),
+                  TimeZone.getDefault()))
+              }
+            }
+            val fullExpectations = (ts ++ impalaExpectations).map {
+              _.toString()
+            }.sorted.toArray
+            val actual = read.map {
+              _.getTimestamp(0).toString()
+            }.sorted
+            withClue(s"applyConversion = $applyConversion; vectorized = $vectorized") {
+              assert(fullExpectations === actual)
+
+              // Now test that the behavior is still correct even with a filter which could get
+              // pushed down into parquet.  We don't need extra handling for pushed down
+              // predicates because (a) in ParquetFilters, we ignore TimestampType and (b) parquet
+              // does not read statistics from int96 fields, as they are unsigned.  See
+              // scalastyle:off line.size.limit
+              // https://github.com/apache/parquet-mr/blob/2fd62ee4d524c270764e9b91dca72e5cf1a005b7/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java#L419
+              // https://github.com/apache/parquet-mr/blob/2fd62ee4d524c270764e9b91dca72e5cf1a005b7/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java#L348
+              // scalastyle:on line.size.limit
+              //
+              // Just to be defensive in case anything ever changes in parquet, this test checks
+              // the assumption on column stats, and also the end-to-end behavior.
+
+              val hadoopConf = sparkContext.hadoopConfiguration
+              val fs = FileSystem.get(hadoopConf)
+              val parts = fs.listStatus(new Path(tableDir.getAbsolutePath), new PathFilter {
+                override def accept(path: Path): Boolean = !path.getName.startsWith("_")
+              })
+              // grab the meta data from the parquet file.  The next section of asserts just make
+              // sure the test is configured correctly.
+              assert(parts.size == 2)
+              parts.map { part =>
+                val oneFooter =
```

Comment:
Alternatively, we could do this as below maybe:

```scala
val hadoopConf = sparkContext.hadoopConfiguration
val footers = readAllFootersWithoutSummaryFiles(
  new Path(tableDir.getAbsolutePath), hadoopConf)
footers.map { footer =>
  val metadata = footer.getParquetMetadata
  ...
```

Reply:
I don't have any particular preference, but just for curiosity, is there any reason to prefer that version instead?

Reply:
I just found it shorter. I am fine.